### PR TITLE
backupccl: shard huge tests into several tests

### PR DESCRIPTION
**Pull request by healthy-pod**
This code change shards the testcases in `TestRestoreMemoryMonitoring`,
`TestRestoreEntryCover`, `TestRestoreMidSchemaChange` into separate tests
and shards test files in `TestDataDriven` into separate tests.

These tests were a bottleneck on CCL Unit Tests time and the sharding results
in an average of 7-11 minutes improvement.

Release note: None
Epic: none

### CODE REVIEW

**Comment by healthy-pod** [comment]
Changes in #103965 also helped ensure `TestFullClusterBackup` will always run in less than 2 minutes.

**Comment by healthy-pod** [comment]
Note: the current bottleneck are some `schemachangerccl` tests that take more than 6 minutes to run (we/sql-foundations should deal with them in another PR)

**Comment by healthy-pod** [comment]
Ready for review

File: pkg/ccl/backupccl/BUILD.bazel [hunk 1]
```diff
@@ -345,4 +350,29 @@ stringer(
     typ = "targetScope",
 )
 
+genrule(
+    name = "gen-backupccl-tests",
+    srcs = [
+        "//pkg/ccl/backupccl:testdata",
+    ],
+    outs = [
+        "restore_memory_monitoring_generated_test.go",
```
Comment (rickystewart) [inline pkg/ccl/backupccl/BUILD.bazel:359]: Nit: alphabetize

File: pkg/ccl/backupccl/restore_mid_schema_change_test.go [hunk 2]
```diff
@@ -53,60 +51,53 @@ import (
...
 // matches the backup name. If the backup is expected to contain several tables,
 // the table names will be backupName1, backupName2, ...
-func TestRestoreMidSchemaChange(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-
-	skip.UnderRaceWithIssue(t, 56584)
-
+//
+//lint:ignore U1000 unused
```
Comment (rickystewart) [inline pkg/ccl/backupccl/restore_mid_schema_change_test.go:55]: I would expect that these functions, being used by generated code, do not trigger the unused checker. Is that not the case?

**Review by rickystewart — APPROVED** [review]

**Comment by healthy-pod** [comment]
bors r=rickystewart

**Comment by mgartner** [comment]
Do the generated files need to be regenerated when adding a new test file? I don't see any instructions for how to run `testgen` and `go run pkg/ccl/backupccl/testgen/main.go` does not work.

**Comment by healthy-pod** [comment]
> Do the generated files need to be regenerated when adding a new test file? I don't see any instructions for how to run testgen and go run pkg/ccl/backupccl/testgen/main.go does not work.

@mgartner These generated files are not checked in. You can generate them using `./dev gen go` but you don't need to unless you want to run one of the generated tests locally.

**Pull request merged**